kernel: gather basic thread runtime statistics

dcpleung opened the pull request:
This adds the bits to gather the first thread runtime statictic:
thread execution time. This follows the tracing subsys to
instrument the point at which threads get switched in and out.
It provides a rough idea of how much time a thread is spent
in active execution.

Note this only adds the instrumentation point in the general
location. Arch-specifc ones to follow in later patches.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: gather basic thread runtime statistics

This adds the bits to gather the first thread runtime statictic:
thread execution time. It provides a rough idea of how much time
a thread is spent in active execution. Currently it is not being
used, pending following commits where it combines with the trace
points on context switch as they instrument the same locations.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
shell: show thread execution cycles

This adds the bits to show the thread execution cycles
when doing "kernel threads" in shell.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

nashif commented:
This shows that we really need a better way to capture timing of context switch, with this, we now have three configs that essentially doing the same thing and that is all in asm code. Maybe what we need is just a call to a c function that is independent of what feature is active now (benchmarking, tracing or runtime stats..) where all of this can happen in C code.

I had the intention to removing the EXECUTION_BENCHMARKING already and make it a special tracing mode for measuring data, however now I realise this will make tracing and benchmarking (and cpu stats that we have right now) all mutually exclusive.

nashif commented:
1134 % :-)
```
 0x110800 idle 00
	options: 0x1, priority: 15 timeout: 0
	state:
	Total execution cycles: 3131838434 (1134 %)
	stack size 256, unused 148, usage 108 / 256 (42 %)

```

dcpleung commented:
> This shows that we really need a better way to capture timing of context switch, with this, we now have three configs that essentially doing the same thing and that is all in asm code. Maybe what we need is just a call to a c function that is independent of what feature is active now (benchmarking, tracing or runtime stats..) where all of this can happen in C code.
> 
> I had the intention to removing the EXECUTION_BENCHMARKING already and make it a special tracing mode for measuring data, however now I realise this will make tracing and benchmarking (and cpu stats that we have right now) all mutually exclusive.

Since there are `z_thread_mark_switched_in/out()`, it is possible to roll the `sys_trace_thread_in/out()` inside them. So it would be easier to manage in the future.

As for the 1134%, it is possibly due to timer overflowing the 64-bit counter. If you are running QEMU with icount, the 64-bit counter wraps around so much you can see it with consecutive `kernel threads` command.

dcpleung committed:
tracing: roll thread switch in/out into thread stats functions

Since the tracing of thread being switched in/out has the same
instrumentation points, we can roll the tracing function calls
into the one for thread stats gathering functions.
This avoids duplicating code to call another function.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

ioannisg commented on kernel/Kconfig:
```diff
@@ -343,6 +343,15 @@ config THREAD_MAX_NAME_LEN
+config THREAD_RUNTIME_STATS
+	bool "Thread runtime statistics"
+	help
+	  Gather thread runtime statistics.
```
@dcpleung @nashif how does this _really_ differ from TRACING CPU_STATS?
I think there should be one way to extract thread run-time measurements (and then also obtain statistics and duty-cycling) and it should be used for Tracing, Benchmarking, etc. What do you think`?

nashif replied:
> how does this _really_ differ from TRACING CPU_STATS?

- TRACING_CPU_STATS is per cpu, this is per thread.
- TRACING_CPU_STATS is actually "mis-using" the tracing feature for capturing timing information, it does not trace anything.
- With TRACING_CPU_STATS enabled, you can't do tracing at the same time, so it is mutually exclusive feature, the above i independent, so you can do tracing and capture runtime stats per thread.
- Once we have this in and satisfies all TRACING_CPU_STATS features, we should consider dropping TRACING_CPU_STATS.

ioannisg replied:
OK, thanks for the clarification.
Yeap, I also had the understanding that CPU_STATS is capturing timing information (for calculating duty-cycling).

So, do we agree that THREAD_RUNTIME_STATS should be the only Kconfig switch for enabling timestamp capturing in ARCHitecture and Kernel functions?
Then THREAD_RUNTIME_STATS should be used for both bench-marking and tracing?

nashif replied:
> Then THREAD_RUNTIME_STATS should be used for both bench-marking and tracing?

those are 2 different things, not sure what you mean here.

nashif commented on arch/arm/core/aarch32/swap_helper.S:
```diff
@@ -51,10 +51,10 @@ GDATA(_kernel)
 SECTION_FUNC(TEXT, z_arm_pendsv)
 
-#ifdef CONFIG_TRACING
+#if defined(CONFIG_TRACING) || defined(CONFIG_THREAD_RUNTIME_STATS)
```
we could get rid of this #if here and everwhere else where this function is called and move things to where the function is implemented. It would be an empty function if no tracing or runtime stats are configured, leaves us with less ifdefs and and someone else decides to use this hook, there will be only once place to change.

dcpleung replied:
My understanding is that an empty function in another file would still get called. So here we need to do push/pop even when tracing or runtime stats are not enabled. I can add an indirect kconfig (where the others would select this one) to replace the current #ifdef. So we don't need to branch/jump unless necessary.

nashif replied:
fair

dcpleung replied:
Replaced with an indirect kconfig.

nashif commented on include/kernel.h:
```diff
@@ -5356,6 +5371,26 @@ __syscall void k_str_out(char *c, size_t n);
+ * @param thread ID of thread.
+ * @return The total execution cycles
+ */
+static inline uint64_t k_thread_total_exec_cycles_get(struct k_thread *thread)
```
ugly name, maybe k_thread_runtime_stats_get()?

dcpleung replied:
My assumption is that the runtime stats will be expanded in the future to gather more metrics. So the get function needs to be specified on what it's returning. Let me think of another name.

nashif replied:
maybe put the data in a struct and return all information available, so we do not need to create an API for every new metric

dcpleung replied:
Done.

nashif commented on include/kernel.h:
```diff
@@ -5356,6 +5371,26 @@ __syscall void k_str_out(char *c, size_t n);
+ *
+ * @return Cumulative execution cycles of all threads
+ */
+uint64_t k_thread_all_exec_cycles_get(void);
```
maybe k_thread_runtime_stats_all_get()

dcpleung replied:
Done.

dcpleung committed:
timing: do not repeatedly do init()/start()/stop()

We should not be initializing/starting/stoping timing functions
multiple times. So this changes how the timing functions are
structured to allow only one initialization, only start when
stopped, and only stop when started.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: use timing functions to gather thread runtime stats

This uses the timing functions to gather execution cycles of
threads. This provides greater details if arch/SoC/board
uses timer with higher resolution.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

nashif commented on kernel/thread.c:
```diff
@@ -995,3 +1003,64 @@ static inline k_ticks_t z_vrfy_k_thread_timeout_expires_ticks(
+	struct k_thread *thread;
+
+	thread = k_current_get();
+	thread->rt_stats.last_switched_in = k_cycle_get_32();
```
shouldnt you be getting the timing information using the timing APIs?

dcpleung replied:
Another commit adds the kconfig to use timing functions. This has the flexibility to not use timing functions as the timer used by timing functions may be reserved for something else in the app.

dcpleung committed:
nios2: fix register saving during thread switch instrumentation

This changes to use stack to store registers before calling thread
switch instrumentation functions, instead of using the thread's
register saving struct.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

nashif commented:
@dcpleung  please rebase.

dcpleung commented:
> @dcpleung please rebase.

Done.

andrewboie commented on arch/x86/core/ia32/swap.S:
```diff
@@ -66,9 +66,9 @@
 SECTION_FUNC(TEXT, arch_swap)
-#if defined(CONFIG_TRACING)
+#if defined(CONFIG_INSTRUMENT_THREAD_SWITCHING)
 	pushl	%eax
```
not strictly part of this PR. but are we using EAX for something here? IA32 passes in all args via stack so maybe this is unnecessary

andrewboie commented on include/kernel.h:
```diff
@@ -5062,6 +5095,28 @@ __syscall void k_str_out(char *c, size_t n);
+ * @param stats Pointer to struct to copy statistics into.
+ * @return -EINVAL if null pointers, otherwise 0
+ */
+int k_thread_runtime_stats_get(struct k_thread *thread,
```
use `k_tid_t` here instead of `struct k_thread *` for application-facing APIs that take thread references.

(yes there is cleanup elsewhere in the kernel with respect to this)

dcpleung replied:
Changed to use `k_tid_t`.

andrewboie commented on include/timing/timing.h:
```diff
@@ -10,15 +10,49 @@
+timing_t arch_timing_counter_get(void);
+uint64_t arch_timing_cycles_get(volatile timing_t *const start,
+				volatile timing_t *const end);
+uint64_t arch_timing_freq_get(void);
```
these prototypes need to be in an arch_interface header and not here, with complete documentation.

dcpleung replied:
Moved these functions to `sys/arch_interface.h`.

tyrantlizard77 reviewed (commented):
only big complain is the `arch_` interfaces need to be in their specific headers with complete doxygen.

I think we need some doc/ level documentation for this feature as well.

dcpleung committed:
doc: add documentation for timing functions

This adds basic documentation on timing functions,
and how to use them.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
doc: add documentation on thread runtime statistics

Adds a section to thread doc on thread runtime statistics
and how to obtain it.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Updated and also added doc about timing functions and thread runtime statistics.
